### 3. [2 hrs.] Use the data provided in the shared file gasoline_use.txt to:

- I,  the index;
- A0, 1; - used for intercept purposes. Can be ignored if fitting also finds intercept. 
- A1, the gasoline ta- A2, the per capita income ($);
- A2, the per capita income ($);
- A3, the number of miles of paved highway (miles);
- A4, the proportion of drivers (%);
- B,  the consumption of gasoline (gallons).

In [49]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [50]:
with open('gasoline_use.txt', 'r') as f:
    lines = f.readlines()[36:]
    #split first line by , and next lines by space
    lines = [line.split(',') if i == 0 else line.split() for i, line in enumerate(lines)]

#convert to dataframe
df = pd.DataFrame(lines[1:], columns=lines[0])
# Consumption of gasoline (millions of gallons)\n to Consumption of gasoline (millions of gallons)
df.columns = [col.strip() for col in df.columns]
print(df.shape)
df.head()

(48, 7)


Index One Gasolin tax (cents per gallon) Average income (dollars)  \
0     1   1                           9.00                     3571   
1     2   1                           9.00                     4092   
2     3   1                           9.00                     3865   
3     4   1                           7.50                     4870   
4     5   1                           8.00                     4399   

  Paved Highways (miles) Proportion of population with driver's licenses  \
0                   1976                                          0.5250   
1                   1250                                          0.5720   
2                   1586                                          0.5800   
3                   2351                                          0.5290   
4                    431                                          0.5440   

  Consumption of gasoline (millions of gallons)  
0                                           541  
1                                           524  
2                                           561  
3                                           414  
4                                           410

In [51]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48 entries, 0 to 47
Data columns (total 7 columns):
 #   Column                                           Non-Null Count  Dtype 
---  ------                                           --------------  ----- 
 0   Index                                            48 non-null     object
 1   One                                              48 non-null     object
 2   Gasolin tax (cents per gallon)                   48 non-null     object
 3   Average income (dollars)                         48 non-null     object
 4   Paved Highways (miles)                           48 non-null     object
 5   Proportion of population with driver's licenses  48 non-null     object
 6   Consumption of gasoline (millions of gallons)    48 non-null     object
dtypes: object(7)
memory usage: 2.8+ KB


In [52]:
# columns 3 - 7 are object type, convert to float
for col in df.columns[1:7]:
    df[col] = df[col].astype(float)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48 entries, 0 to 47
Data columns (total 7 columns):
 #   Column                                           Non-Null Count  Dtype  
---  ------                                           --------------  -----  
 0   Index                                            48 non-null     object 
 1   One                                              48 non-null     float64
 2   Gasolin tax (cents per gallon)                   48 non-null     float64
 3   Average income (dollars)                         48 non-null     float64
 4   Paved Highways (miles)                           48 non-null     float64
 5   Proportion of population with driver's licenses  48 non-null     float64
 6   Consumption of gasoline (millions of gallons)    48 non-null     float64
dtypes: float64(6), object(1)
memory usage: 2.8+ KB


#### 3.1. [10 points.] Show the equation found by fitting the training data:
**y = f(x1, x2, x3, x4) = a0+ a1x1 + a2x2 + a3x3 + a4x4**

In [53]:
# linear regression
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

In [54]:
#split data into train and test
x = df.iloc[:, 1:6]
y = df.iloc[:, 6]
x_train, x_test, y_train, y_test = train_test_split(x, y, random_state=0)

# fit the model
model = LinearRegression()
model.fit(x_train, y_train)

# predict
y_pred = model.predict(x_test)

# show the equation
print('y = {:.2f} {:+.2f}X1 {:+.2f}X2 {:+.2f}X3 {:+.2f}X4 {:+.2f}'.format(*model.coef_, model.intercept_))

y = 0.00 -42.74X1 -0.07X2 -0.01X3 +1329.99X4 +487.37


#### 3.2. [5 points.] What is the R2 for the prediction of y? Use testing data.

In [59]:
# find the R^2
from sklearn.metrics import r2_score
#print('R2 square of the training data: ', r2_score(y_train, model.predict(x_train)))
print('R2 square of the test data: ', r2_score(y_test, y_pred))

R2 square of the test data:  0.16216522232680952


#### 3.3. [5 points.] What would happen to gasoline consumption if taxes are increased by $2.00? Use training data.

In [56]:
# plus 200.00 in every line Gasolin tax (cents per gallon)
df['Gasolin tax (cents per gallon)'] = df['Gasolin tax (cents per gallon)'] + 200
df.head()

Index  One  Gasolin tax (cents per gallon)  Average income (dollars)  \
0     1  1.0                           209.0                    3571.0   
1     2  1.0                           209.0                    4092.0   
2     3  1.0                           209.0                    3865.0   
3     4  1.0                           207.5                    4870.0   
4     5  1.0                           208.0                    4399.0   

   Paved Highways (miles)  Proportion of population with driver's licenses  \
0                  1976.0                                            0.525   
1                  1250.0                                            0.572   
2                  1586.0                                            0.580   
3                  2351.0                                            0.529   
4                   431.0                                            0.544   

   Consumption of gasoline (millions of gallons)  
0                                          541.0  
1                                          524.0  
2                                          561.0  
3                                          414.0  
4                                          410.0

In [57]:
#split data into train and test
# all columns except the last one and 3rd column
x = df.iloc[:, 1:6]
y = df.iloc[:, 6]
x_train, x_test, y_train, y_test = train_test_split(x, y, random_state=0)

# fit the model
model = LinearRegression()
model.fit(x_train, y_train)

# predict
y_pred = model.predict(x_test)

# show the equation
print('y = {:.2f} {:+.2f}X1 {:+.2f}X2 {:+.2f}X3 {:+.2f}X4 {:+.2f}'.format(*model.coef_, model.intercept_))

y = 0.00 -42.74X1 -0.07X2 -0.01X3 +1329.99X4 +9034.72


In [58]:
# find the R^2
from sklearn.metrics import r2_score
print('R2 square of the training data: ', r2_score(y_train, model.predict(x_train)))
print('R2 square of the test data: ', r2_score(y_test, y_pred))

R2 square of the training data:  0.7202870624476543
R2 square of the test data:  0.16216522232680952


**ANSWER:** A intecept has changed from 487.37 to 9034.72, and another is not changed.